### DISCUSSION TITLE
笔记模板分享[Item] 创新卡

### DISCUSSION THREAD

**Discussion by liuenqian** [windingwind/zotero-better-notes#1072] [Note Templates]
### Description

可以对高亮笔记按照嵌套标签或者标签进行生成模板笔记。感谢[MuiseDestiny](https://github.com/MuiseDestiny)帮忙修改的代码。

### Screenshots

![image](https://github.com/user-attachments/assets/8c9adf99-42b4-4798-b279-c912fff192fe)

### Test on Zotero version

7.0.0-beta.108+80cc97125

### Test on Better Notes version

1.1.4-beta.111

### Template Type

Item

### Template Share Code

```yaml
# This template is specifically for importing/sharing, using better 
# notes 'import from clipboard': copy the content and
# goto Zotero menu bar, click Edit->New Template from Clipboard.  
# Do not copy-paste this to better notes template editor directly.
name: "[Item] 创新卡"
content: |-
  ${await new Promise(async (r) => {
   const getAnnotationsByTag = async (tagName) => {
    const pdfItem = await topItem.getBestAttachment();
    let result = ""
    for (let annoItem of pdfItem.getAnnotations()) {
     if (annoItem.getTags().find(i=>i.tag.includes(tagName))) {
  
      const res = Zotero.EditorInstanceUtilities.serializeAnnotations(
       [
        {
        ...(await Zotero.Annotations.toJSON(annoItem)), ...{ attachmentItemID: annoItem.parentID }
        }
       ]
      )
      result += res.html
     }
    }
    return result
   }
      let res = `<h1>创新卡 - 🔤${topItem.getField("extra").match(/titleTranslation:(.+)/)[1]} (<span style="background-color: #B1A5E1">${topItem.getField("year")}</span>, <span style="background-color: #CBE1A5">${topItem.getCreators().map((v)=>v.lastName).slice(0,1)}</span>) ${topItem.getField("journalAbbreviation")} (${topItem.getField("libraryCatalog")}) </h1>
  <p><strong><span style="color: #283593">原名：</span></strong>${topItem.getField('title')}</p >
  <p><strong><span style="color: #283593">译名：</span></strong>${topItem.getField("extra").match(/titleTranslation:(.+)/)[1]} </p >
  <p><strong><span style="color: #283593">作者：</span></strong>${topItem.getCreators().map((v)=>v.lastName).slice(0,1)} et al.</p >
  <p><strong><span style="color: #283593">期刊：</span></strong>${topItem.getField('publicationTitle')}</p >
  <p><strong><span style="color: #283593">IFQ：</span></strong>${{
  let space = " ㅤㅤ ㅤㅤ"
  return Array.prototype.map.call(
   Zotero.ZoteroStyle.api.renderCell(topItem, "publicationTags").childNodes,
   e => {
    e.innerText =  space + e.innerText + space;
    return e.outerHTML
   }
   ).join(space)
  }}$</p >
  <p><strong><span style="color: #283593">DOI：</span></strong>${topItem.getField('DOI')}</p >
  <p><strong><span style="color: #283593">发表时间：</span></strong>${topItem.getField('date')}</p >
  <h2>👉创新点</h2>
  <h3><span style="color: #063C89">${topItem.getField('journalAbbreviation')} (IF = ${topItem.getField('libraryCatalog')}): ${topItem.getField("extra").match(/titleTranslation:(.+)/)[1]}；${topItem.getField('date')}</span></h3>
  <h4><span style="color: #A61B41">局限性：</span></h4>
  <p> ${await getAnnotationsByTag('背景/局限性')} </p >
  
  <h4><span style="color: #A61B41">创新点：</span></h4>
  <p>${await getAnnotationsByTag('创新点')}</p >
  
  <h4><span style="color: #A61B41">研究假设：</span></h4>
  <p>${await getAnnotationsByTag('背景/研究假设')}</p >
  
  <h4><span style="color: #A61B41">材料：</span></h4>
  <p></p >
  
  <h4><span style="color: #A61B41">方法：</span></h4>
  <p>${await getAnnotationsByTag('指标/材料实验/材料合成')}</p >
  
  <h4><span style="color: #A61B41">结果：</span></h4>
  <p>${await getAnnotationsByTag('结果/材料合成')}</p >
  
  <p> <span style="color: #063C89">原文链接：</span></p >
   <a href="${topItem.getField('url')}" target="_blank"> ${topItem.getField('url')}</a>
  
  
  `;
      r(res);
    })}
```

**Comment by huihuang-chn** [comment]
![1726931299613](https://github.com/user-attachments/assets/5c78a765-6609-41f9-b991-7e94b0de804a)
如图，直接复制以后，生成模板，无法用，笔记空白，请检查一下代码，我是小白看不出来。加油

**Reply by Beibeihai1** [reply]
19行  大括号后面加分号;

**Reply by huihuang-chn** [reply]
加了，不过还是不行，不知道啥原因？你可以复制模板试一下，预览那里也不显示，在条目直接添加，笔记界面是空白。

**Comment by huihuang-chn** [comment]
&nbsp; ${await new Promise(async (r) =&gt; {
&nbsp; &nbsp;const getAnnotationsByTag = async (tagName) =&gt; {
&nbsp; &nbsp; const pdfItem = await topItem.getBestAttachment();
&nbsp; &nbsp; let result = ""
&nbsp; &nbsp; for (let annoItem of pdfItem.getAnnotations()) {
&nbsp; &nbsp; &nbsp;if (annoItem.getTags().find(i=&gt;i.tag.includes(tagName))) {
&nbsp; 
&nbsp; &nbsp; &nbsp; const res = Zotero.EditorInstanceUtilities.serializeAnnotations(
&nbsp; &nbsp; &nbsp; &nbsp;[
&nbsp; &nbsp; &nbsp; &nbsp; {
&nbsp; &nbsp; &nbsp; &nbsp; ...(await Zotero.Annotations.toJSON(annoItem)), ...{ attachmentItemID: annoItem.parentID }
&nbsp; &nbsp; &nbsp; &nbsp; }
&nbsp; &nbsp; &nbsp; &nbsp;]
&nbsp; &nbsp; &nbsp; )
&nbsp; &nbsp; &nbsp; result += res.html
&nbsp; &nbsp; &nbsp;}
&nbsp; &nbsp; }
&nbsp; &nbsp; return result
&nbsp; &nbsp;};
&nbsp; &nbsp; &nbsp; let res = `<h1&gt;创新卡 - 🔤${topItem.getField("extra").match(/titleTranslation:(.+)/)[1]} (<span style="background-color: #B1A5E1"&gt;${topItem.getField("year")}</span&gt;, <span style="background-color: #CBE1A5"&gt;${topItem.getCreators().map((v)=&gt;v.lastName).slice(0,1)}</span&gt;) ${topItem.getField("journalAbbreviation")} (${topItem.getField("libraryCatalog")}) </h1&gt;
&nbsp; <p&gt;<strong&gt;<span style="color: #283593"&gt;原名：</span&gt;</strong&gt;${topItem.getField('title')}</p &gt;
&nbsp; <p&gt;<strong&gt;<span style="color: #283593"&gt;译名：</span&gt;</strong&gt;${topItem.getField("extra").match(/titleTranslation:(.+)/)[1]} </p &gt;
&nbsp; <p&gt;<strong&gt;<span style="color: #283593"&gt;作者：</span&gt;</strong&gt;${topItem.getCreators().map((v)=&gt;v.lastName).slice(0,1)} et al.</p &gt;
&nbsp; <p&gt;<strong&gt;<span style="color: #283593"&gt;期刊：</span&gt;</strong&gt;${topItem.getField('publicationTitle')}</p &gt;
&nbsp; <p&gt;<strong&gt;<span style="color: #283593"&gt;IFQ：</span&gt;</strong&gt;${{
&nbsp; let space = " ㅤㅤ ㅤㅤ"
&nbsp; return Array.prototype.map.call(
&nbsp; &nbsp;Zotero.ZoteroStyle.api.renderCell(topItem, "publicationTags").childNodes,
&nbsp; &nbsp;e =&gt; {
&nbsp; &nbsp; e.innerText = &nbsp;space + e.innerText + space;
&nbsp; &nbsp; return e.outerHTML
&nbsp; &nbsp;}
&nbsp; &nbsp;).join(space)
&nbsp; }}$</p &gt;
&nbsp; <p&gt;<strong&gt;<span style="color: #283593"&gt;DOI：</span&gt;</strong&gt;${topItem.getField('DOI')}</p &gt;
&nbsp; <p&gt;<strong&gt;<span style="color: #283593"&gt;发表时间：</span&gt;</strong&gt;${topItem.getField('date')}</p &gt;
&nbsp; <h2&gt;👉创新点</h2&gt;
&nbsp; <h3&gt;<span style="color: #063C89"&gt;${topItem.getField('journalAbbreviation')} (IF = ${topItem.getField('libraryCatalog')}): ${topItem.getField("extra").match(/titleTranslation:(.+)/)[1]}；${topItem.getField('date')}</span&gt;</h3&gt;
&nbsp; <h4&gt;<span style="color: #A61B41"&gt;局限性：</span&gt;</h4&gt;
&nbsp; <p&gt; ${await getAnnotationsByTag('背景/局限性')} </p &gt;
&nbsp; 
&nbsp; <h4&gt;<span style="color: #A61B41"&gt;创新点：</span&gt;</h4&gt;
&nbsp; <p&gt;${await getAnnotationsByTag('创新点')}</p &gt;
&nbsp; 
&nbsp; <h4&gt;<span style="color: #A61B41"&gt;研究假设：</span&gt;</h4&gt;
&nbsp; <p&gt;${await getAnnotationsByTag('背景/研究假设')}</p &gt;
&nbsp; 
&nbsp; <h4&gt;<span style="color: #A61B41"&gt;材料：</span&gt;</h4&gt;
&nbsp; <p&gt;</p &gt;
&nbsp; 
&nbsp; <h4&gt;<span style="color: #A61B41"&gt;方法：</span&gt;</h4&gt;
&nbsp; <p&gt;${await getAnnotationsByTag('指标/材料实验/材料合成')}</p &gt;
&nbsp; 
&nbsp; <h4&gt;<span style="color: #A61B41"&gt;结果：</span&gt;</h4&gt;
&nbsp; <p&gt;${await getAnnotationsByTag('结果/材料合成')}</p &gt;
&nbsp; 
&nbsp; <p&gt; <span style="color: #063C89"&gt;原文链接：</span&gt;</p &gt;
&nbsp; &nbsp;<a href="${topItem.getField('url')}" target="_blank"&gt; ${topItem.getField('url')}</a&gt;
&nbsp; 
&nbsp; 
&nbsp; `;
&nbsp; &nbsp; &nbsp; r(res);
&nbsp; &nbsp; })}











【黑色幽默】
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "windingwind/zotero-better-notes"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年10月3日(星期四) 晚上11:07
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [windingwind/zotero-better-notes] 笔记模板分享[Item] 创新卡 (Discussion #1072)





 
19行  大括号后面加分号;

**Comment by lilyomo** [comment]
你好，请教下这个模板该如何使用，没搞明白。是能够将高亮文本自动添加到笔记中吗？

**Reply by huihuang-chn** [reply]
看截图应该是，不过代码复制过去以后显示错误，不能用
